src/Database/LSMTree/Internal/WriteBufferBlobs.hs

```diff
@@ -0,0 +1,262 @@
+-- runs (and the write buffer) differently: avoiding copying when blobs are
+-- merged and using some kind of GC algorithm to recover space for blobs that
+-- are not longer needed. There are LSM algorithms that do this for values
+-- (i.e. copying keys only during merge and referring to values managed in a
+-- separate disk heap), so the same could be applied to blobs.
+--
+data WriteBufferBlobs m h =
+     WriteBufferBlobs {
+       blobFileState      :: !(MutVar (PrimState m) (BlobFileState m h))
+
+       -- | The name is reserved in advance, even if it is not used.
+     , blobFileName       :: !FsPath
+
+       -- | The reference counter for the blob file.
+     , blobFileRefCounter :: {-# UNPACK #-} !(RC.RefCounter m)
+     }
+
+-- | The blob file itself is opened on-demand, to avoid the cost in cases where
+-- blobs are not needed.
+data BlobFileState m h =
+
+     NoBlobFileYet
+   | OpenBlobFile {
+       blobFileHandle  :: {-# UNPACK #-} !(FS.Handle h)
+
+       -- | The manually tracked file pointer.
+     , blobFilePointer :: !(FilePointer m)
+     }
+
+new :: PrimMonad m
+    => HasFS m h
+    -> FS.FsPath
+    -> m (WriteBufferBlobs m h)
+new fs blobFileName = do
+    blobFileState <- P.newMutVar NoBlobFileYet
+    blobFileRefCounter <- RC.mkRefCounter1 (Just (finaliser fs blobFileState))
+    return WriteBufferBlobs {
+      blobFileState,
+      blobFileName,
+      blobFileRefCounter
+    }
+
+finaliser :: PrimMonad m
+          => HasFS m h
+          -> MutVar (PrimState m) (BlobFileState m h)
+          -> m ()
+finaliser fs bf = do
+    bfs <- P.readMutVar bf
+    case bfs of
+      OpenBlobFile {blobFileHandle} -> FS.hClose fs blobFileHandle
+      NoBlobFileYet {}              -> return ()
+
+addReference :: PrimMonad m => WriteBufferBlobs m h -> m ()
+addReference WriteBufferBlobs {blobFileRefCounter} =
+    RC.addReference blobFileRefCounter
+
+removeReference :: (PrimMonad m, MonadMask m) => WriteBufferBlobs m h -> m ()
+removeReference WriteBufferBlobs {blobFileRefCounter} =
+    RC.removeReference blobFileRefCounter
+
+addBlob :: (PrimMonad m, MonadThrow m)
+        => HasFS m h
+        -> WriteBufferBlobs m h
+        -> SerialisedBlob
+        -> m BlobSpan
+addBlob fs WriteBufferBlobs {blobFileName, blobFileState} blob = do
+    bfs <- P.readMutVar blobFileState
+    case bfs of
+      OpenBlobFile {blobFileHandle, blobFilePointer} ->
+        writeBlob blobFileHandle blobFilePointer
+
+      NoBlobFileYet -> do
+        -- Must use read/write mode because we write blobs when adding, but
+        -- we can also be asked to retrieve blobs at any time.
+        blobFileHandle <- FS.hOpen fs blobFileName (FS.ReadWriteMode FS.MustBeNew)
+        blobFilePointer <- newFilePointer
+        P.writeMutVar blobFileState OpenBlobFile {
+          blobFileHandle,
+          blobFilePointer
+        }
+        writeBlob blobFileHandle blobFilePointer
```

Comment:
This is a good point. It _is_ intended to be used concurrently, since it'll be shared between many table handles (when table handles are duplicated). And any of the table handles could then be the first to write a blob. So this bit could be used concurrently.

I had been thinking that this bit could only happen with the original creator, but that's not true, we could duplicate and then add.

I wonder if it'd be better just to open the file eagerly. It'd eliminate this situation, and make blob refs easier to construct.

Reply:
> I wonder if it'd be better just to open the file eagerly. It'd eliminate this situation, and make blob refs easier to construct.

I don't have a strong opinion -- though if it's a trade off between simpler code and redundant, though infrequent, file operations, maybe we should go for code clarity

Reply:
I'm doing the simplification.